### PR TITLE
[MRG] Add Pillow's tutorial for RGC spikes prediction

### PR DESCRIPTION
**Pull request by titipata**
This pull request contains a tutorial for linear Gaussian and Poisson GLM. This PR need the dataset from https://github.com/glm-tools/datasets/pull/1.

### CODE REVIEW

**Comment by jasmainak** [comment]
you have conflicts

File: examples/plot_glm_retinal_ganglia.py [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+for sharing their tutorial dataset. The original MATLAB and Python tutorial can be found from
+https://github.com/pillowlab/GLMspiketraintutorial.
+
+These data were collected by Valerie Uzzell in the lab of
+E.J. Chichilnisky at the Salk Institute.  For full information see
+Uzzell et al. (J Neurophys 04), or Pillow et al. (J Neurosci 2005).
+**Again, the dataset in this tutorial is granted by the original authors
+for educational use only not for publication**. If you want to use it
+beyond the educational purposes,
+please contact ``pillow@princeton.edu``
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: make this a prompt as I commented in the other PR :)

File: examples/plot_glm_retinal_ganglia.py [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+The dataset contains spike responses from 2 ON and 2 OFF parasol
+retinal ganglion cells (RGCs) in primate retina, stimulated with
+full-field `binary white noise`. Two experiment performed consisted of a
+long (20-minute) binary stochastic (non-repeating) stimulus
+which can be used for computing the spike-triggered average
+(or characterizing some other model of the response).
+"""
+
+# Authors: Jonathan Pillow <pillow@princeton.edu>
+# Edited by: Titipat Achakulvisut <my.titipat@gmail.com>, Konrad Kording <koerding@gmail.com>
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: just list yourself as authors, no need to say "edited by". We are in open source world.

File: examples/plot_glm_retinal_ganglia.py [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+* The dataset is granted by the original authors for the educational use in this tutorial
+* If you want to use it beyond educational purposes, please contact ``pillow@princeton.edu``
+
+First of all, we would like to thank Professor Johnathan Pillow and Professor E.J. Chichilnisky
+for sharing their tutorial dataset. The original MATLAB and Python tutorial can be found from
+https://github.com/pillowlab/GLMspiketraintutorial.
+
+These data were collected by Valerie Uzzell in the lab of
+E.J. Chichilnisky at the Salk Institute.  For full information see
+Uzzell et al. (J Neurophys 04), or Pillow et al. (J Neurosci 2005).
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: you can add a proper reference. See [how we do it in mne-bids](https://github.com/mne-tools/mne-bids/blob/master/examples/convert_group_studies.py#L13)

File: examples/plot_glm_retinal_ganglia.py [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+         color='green', linewidth=2, label='poissonGLM')
+plt.plot(t_sample, spikes_pred_poissonGLM_coupled[sample_index],
+         color='pink', linewidth=2, label='poissonGLM_couple')
+
+plt.xlim([t_sample.min(), t_sample.max()])
+plt.title('Spike counts prediction with spikes couple using Poisson GLM')
+plt.ylabel('Binned Spike Counts')
+plt.legend()
+plt.show()
+print('Training perf using Pyglmnet score {:.2f}'.\
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: This example is way too long. We can't merge it unless it's less than 100 lines :) Can you aim to do that?

**Comment by jasmainak** [comment]
@titipata can you point the URL to your PR in the dataset repo so we can see how the example reads in circle CI?

File: examples/plot_glm_retinal_ganglia.py [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+    np.array(rgcs_dataset['spike_times']['cell_0']),
+    np.array(rgcs_dataset['spike_times']['cell_1']),
+    np.array(rgcs_dataset['spike_times']['cell_2']),
+    np.array(rgcs_dataset['spike_times']['cell_3']),
+]
+
+n_cells = len(spike_times) # total number of cells
+dt = stim_times[1] - stim_times[0] # time between the stimulation
+n_t = len(stim) # total number of the stimulation
+f = 1 / dt # frequency of the stimulation
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]:
```suggestion
f = 1. / dt # frequency of the stimulation
```

if `dt` is not a float, you might have a problem. Better be safe

File: examples/plot_glm_retinal_ganglia.py [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+
+########################################################
+#
+# You can pick a cell to work with and visualize the spikes for one second.
+# In this case, we will pick cell number 2 (ON cell).
+
+cell_num = 2 # pick cell number 2 (ON cell)
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]:
```suggestion
cell_idx = 2 # pick cell number 2 (ON cell)
```

**Comment by jasmainak** [comment]
@titipata can you ping me once you made the example shorter. We already have one neuro example that's very hard to follow ... can't do the same again.

File: examples/plot_glm_retinal_ganglia.py (hunk 1 above)
Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: Thanks! Can you put a reference to this from MNE @jasmainak?

Comment (jasmainak) [reply examples/plot_glm_retinal_ganglia.py]: here you go: https://github.com/mne-tools/mne-python/blob/master/mne/datasets/utils.py#L379-L386

Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: Done! Let me know if it looks good to you!

File: examples/plot_glm_retinal_ganglia.py (hunk 3 above)
Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: Made a proper one now. Let me know if it looks good!

File: examples/plot_glm_retinal_ganglia.py (hunk 4 above)
Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: The code part itself is less than 100 lines. I don't think it can be further trimmed while keeping it good quality.

**Comment by jasmainak** [comment]
Did you try running the example using "make html" ? I get:

```py
Traceback (most recent call last):
  File "/Users/mainak/anaconda3/lib/python3.7/site-packages/sphinx_gallery/gen_rst.py", line 480, in _memory_usage
    out = func()
  File "/Users/mainak/anaconda3/lib/python3.7/site-packages/sphinx_gallery/gen_rst.py", line 465, in __call__
    exec(self.code, self.globals)
  File "/Users/mainak/Documents/github_repos/pyglmnet/examples/plot_glm_retinal_ganglia.py", line 64, in <module>
    from pyglmnet.datasets import fetch_RGCs_data
ImportError: cannot import name 'fetch_RGCs_data' from 'pyglmnet.datasets' (/Users/mainak/.local/lib/python3.7/site-packages/pyglmnet/datasets.py)
```

**Comment by titipata** [comment]
@jasmainak I did fix the issue by merging to `master` (probably not the best way). However, I added a license to `datasets.py`.  The tutorial should now pass `flake8` and trimmed down.

**Comment by titipata** [comment]
@jasmainak this PR should be pretty much ready on my side. Not sure where can I get access to the built documentation page on CircleCI.

**Comment by jasmainak** [comment]
@titipata can you fix the merge conflict by rebase? If you're not confident, I can do it for you. I will also try to do a pass over the example to simplify it. I'm a bit busy today though, so I can get to it only tomorrow or Saturday.

Here's the circleCI build from your last successful build: https://518-55302570-gh.circle-artifacts.com/0/html/auto_examples/plot_glm_retinal_ganglia.html#sphx-glr-auto-examples-plot-glm-retinal-ganglia-py

File: examples/plot_glm_retinal_ganglia.py [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+.. [1] Uzzell, V. J., and E. J. Chichilnisky. "Precision of
+   spike trains in primate retinal ganglion cells."
+   Journal of Neurophysiology 92.2 (2004)
+.. [2] Pillow, Jonathan W., et al. "Prediction and decoding of
+   retinal ganglion cell responses with a probabilistic
+   spiking model." Journal of Neuroscience 25.47 (2005)
+
+"""
+
+# Authors: Titipat Achakulvisut <my.titipat@gmail.com>
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: Sorry if I wasn't clear, you can add the original author names here. Only I don't think you need to specify that you edited it because the example would have changed substantially.

File: examples/plot_glm_retinal_ganglia.py [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+# :math:`y \sim \text{Poiss}(\vec{\theta} \cdot \vec{x})`.
+# We call :math:`f^{-1}` a "link function".
+# Here, we can use Pyglmnet's `GLM` to predict the parameters.
+
+# create possion GLM instance
+glm_poisson = GLM(distr='poisson',
+                  verbose=False, alpha=0.05,
+                  max_iter=1000, learning_rate=0.2,
+                  score_metric='pseudo_R2',
+                  reg_lambda=1e-7, eta=4.0)
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: can you use `GLMCV` instead so it's not needed to specify `reg_lambda`?

File: examples/plot_glm_retinal_ganglia.py [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+    np.array(rgcs_dataset['spike_times']['cell_0']),
+    np.array(rgcs_dataset['spike_times']['cell_1']),
+    np.array(rgcs_dataset['spike_times']['cell_2']),
+    np.array(rgcs_dataset['spike_times']['cell_3']),
+]
+
+n_cells = len(spike_times)  # total number of cells
+dt = stim_times[1] - stim_times[0]  # time between the stimulation
+n_t = len(stim)  # total number of the stimulation
+f = 1. / dt  # frequency of the stimulation
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: single letter variable names are not good, you will have trouble doing a Ctrl + F on them. Plus they are not very descriptive.

```suggestion
sfreq = 1. / dt  # frequency of the stimulation
```

File: examples/plot_glm_retinal_ganglia.py [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+           interpolation='nearest')
+plt.xlabel('lags before spike time')
+plt.ylabel('time bin of response')
+plt.title('Sample first 50 rows of design'
+          ' matrix created using Hankel')
+plt.colorbar()
+plt.show()
+
+########################################################
+# **Compute Spike-Triggered Average (STA)**
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: This is very neuro stuff and not used later on in the tutorial, can we remove this?

File: examples/plot_glm_retinal_ganglia.py [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+plt.show()
+
+########################################################
+# **Fitting and predicting with a linear-Gaussian GLM**
+#
+# For a general linear model, an observed spikes can be
+# thought of an underlying parameter
+# :math:`\theta` that control the spike output:
+# :math:`y = \vec{\theta} \cdot \vec{x} + \epsilon` and
+# :math:`y \sim \text{Poiss}(\vec{\theta} \cdot \vec{x})`
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: could we perhaps follow the same notation as in the tutorial?

File: examples/plot_glm_retinal_ganglia.py (hunk 7 above)
Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: Sounds good! I will change it to the original author here.

File: examples/plot_glm_retinal_ganglia.py [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+and Professor E.J. Chichilnisky for sharing their tutorial dataset.
+The original MATLAB and Python tutorial can be found from
+https://github.com/pillowlab/GLMspiketraintutorial.
+
+These data were collected by Valerie Uzzell in the lab of
+E.J. Chichilnisky at the Salk Institute. For full information see
+Uzzell et al. (J Neurophys 04) [1]_, or Pillow et al. (J Neurosci 2005) [2]_.
+
+In this tutorial, we will demonstrate how to fit linear GLM and
+linear-nonlinear GLM (a.k.a Poisson GLM) to predict the spike counts
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: just say Poisson GLM, no? Not to confuse people with terminology

```suggestion
Poisson GLM to predict the spike counts
```

also I didn't quite get the point of this comparison? Is it show that one of them performs better?

File: examples/plot_glm_retinal_ganglia.py [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+
+plt.imshow(Xdsgn[:50, :],
+           cmap='binary',
+           aspect='auto',
+           interpolation='nearest')
+plt.xlabel('lags before spike time')
+plt.ylabel('time bin of response')
+plt.title('Sample first 50 rows of design'
+          ' matrix created using Hankel')
+plt.colorbar()
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: I would actually turn off the colorbar since there are only 3 values. It's a bit confusing.

Comment (jasmainak) [reply examples/plot_glm_retinal_ganglia.py]: or change it somehow to not show a continuous colorbar

**Comment by titipata** [comment]
@jasmainak  will resolve your comments and then let you help with the rebase!

File: examples/plot_glm_retinal_ganglia.py [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+t_lag = dt * np.arange(-n_t_filt + 1, 1)  # time bins for STA (in seconds)
+plt.plot(t_lag, t_lag * 0, 'k--')
+plt.plot(t_lag, sta, 'bo-')
+plt.title('Spike-Triggered Average (STA)')
+plt.xlabel('Time before spike (sec)')
+plt.xlim([t_lag.min(), t_lag.max()])
+plt.show()
+
+########################################################
+# **Fitting and predicting with a linear-Gaussian GLM**
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: just Gaussian?
```suggestion
# **Fitting and predicting with a Gaussian GLM**
```

File: examples/plot_glm_retinal_ganglia.py [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+# np.exp(glm_poisson.beta0_ + X.dot(glm_poisson.beta_))
+spikes_pred_poissonGLM = glm_poisson.predict(Xdsgn)
+
+#############################################################################
+# **Adding spikes history for predicting spike counts**
+#
+# We can even further predict the spikes by using the spikes history.
+# **Note** the spike-history portion of the design
+# matrix had better be shifted so that we aren't allowed to use the spike
+# count on this time bin to predict itself!
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: I think this need to be introduced earlier when you make the hankel matrix. Otherwise it's not clear why you are doing it.

Comment (jasmainak) [reply examples/plot_glm_retinal_ganglia.py]: I mean the text of doing prediction from spike history. The code can stay here.

File: examples/plot_glm_retinal_ganglia.py [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+         [1] * len(spike_time[sample_index]))
+plt.xlim([t_sample.min(), t_sample.max()])
+plt.ylim([0, 2])
+plt.xlabel('Time (s)')
+plt.ylabel('Spikes (0 or 1)')
+plt.show()
+
+########################################################
+#
+# We can use ``scipy``'s function ``hankel`` to make our design matrix.
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: how does cross-validation work with this kind of design matrix? It looks like you are using some samples in your data to predict others?

File: pyglmnet/datasets.py [hunk 17]
```diff
@@ -11,6 +11,24 @@
...
+
+This tutorial dataset (RGC spikes data) is granted by
+original authors (E.J. Chichilnisky and Jonathan Pillow).
+The dataset ramains a property of the original authors.
+Its use and transfer outside tutorial, e.g. for research purposes,
+is prohibited without written consent to the original authors.
+
+If you reference this dataset in your publications, please:
+    1) acknowledge the authors: E.J. Chichilnisky and Jonathan Pillow
+    2) cite thier publications as indicated on the tutorial
```
Comment (jasmainak) [inline pyglmnet/datasets.py]:
```suggestion
    2) cite the publications as indicated in the tutorial
```

**Comment by jasmainak** [comment]
okay I'm done with my pass!

File: examples/plot_glm_retinal_ganglia.py (hunk 12 above)
Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: Yes, the original purpose is to show which one performs better by measure R^2 .

File: examples/plot_glm_retinal_ganglia.py (hunk 11 above)
Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: Made changes according to `pyglmnet` tutorial

File: examples/plot_glm_retinal_ganglia.py (hunk 8 above)
Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: I did try `GLMCV` but it's extremely slow so I just stick with `GLM` for now.

File: examples/plot_glm_retinal_ganglia.py (hunk 15 above)
Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: Added some explanation on how to create a design matrix here.

File: examples/plot_glm_retinal_ganglia.py (hunk 16 above)
Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: Yeah, it's a time series prediction using previous stimulation and spikes history. I don't think you can simply use traditional CV.

**Comment by titipata** [comment]
@jasmainak done with my pass! It would be great if you can help with `rebase` and minor text changes  (or otherwise just tag me for that). :)

File: examples/plot_glm_retinal_ganglia.py [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+# For a general linear model, an observed spikes can be
+# thought of an underlying parameter
+# :math:`\beta_0, \beta` that control the spiking:
+# :math:`y = (\beta_0 + X \beta) \ + \epsilon` and
+# :math:`y \sim \text{Poiss}(\beta_0 + X \beta)` where
+# :math:`X` is the stimulation and history of stimulation
+#
+# We can add an offset or a "constant" to our design matrix.
+# This can be done by concatenating a column of 1 to
+# our previously created design matrix.
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: why do you need to do this? `fit_intercept` does not do the job?

File: examples/plot_glm_retinal_ganglia.py [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+Xstim = hankel(stim_padded[:-n_t_filt + 1], stim[-n_t_filt:])
+Xspikes = hankel(spikes_padded[:-n_t_hist], stim[-n_t_hist:])
+Xdsgn_hist = np.hstack((Xstim, Xspikes))  # design matrix with spikes history
+
+# Now, we are ready to fit Poisson GLM with spikes history.
+# create possion GLM instance
+glm_poisson_hist = GLM(distr='poisson', alpha=0.05,
+                       learning_rate=1.0,
+                       score_metric='pseudo_R2',
+                       reg_lambda=1e-7)
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: @titipata I simplified, the more we stick to defaults, the better.

File: examples/plot_glm_retinal_ganglia.py [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+from tempfile import TemporaryDirectory
+
+########################################################
+#
+# Fetch the dataset. The JSON file contains the
+# following keys:  ``stim`` (binary stochastic stimulation),
+# ``stim_times`` (time of the stimulation), and
+# ``spike_times`` (recorded time of the spikes)
+
+from pyglmnet.datasets import fetch_RGCs_data
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]:
```suggestion
from pyglmnet.datasets import fetch_rgc_data
```

File: examples/plot_glm_retinal_ganglia.py [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+plt.xlim([0., tmax])
+plt.title('Spike count prediction')
+plt.xlabel('Time (sec)')
+plt.ylabel('Binned Spike Counts')
+plt.legend()
+plt.show()
+
+# performance of the fitted models
+mse_lgGLM_offset = np.mean((
+    spikes_binned - spikes_pred_lgGLM_offset) ** 2)
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: can we use `glm.score` here?

File: examples/plot_glm_retinal_ganglia.py [hunk 22]
```diff
@@ -0,0 +1,270 @@
...
+.. [1] Uzzell, V. J., and E. J. Chichilnisky. "Precision of
+   spike trains in primate retinal ganglion cells."
+   Journal of Neurophysiology 92.2 (2004)
+.. [2] Pillow, Jonathan W., et al. "Prediction and decoding of
+   retinal ganglion cell responses with a probabilistic
+   spiking model." Journal of Neuroscience 25.47 (2005)
+
+"""
+
+# Authors: Jonathan Pillow <pillow@princeton.edu>
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: don't forget to add your name!

File: examples/plot_glm_retinal_ganglia.py [hunk 23]
```diff
@@ -0,0 +1,270 @@
...
+    dpath = fetch_RGCs_data(dpath=temp_dir, accept_rgcs_license=True)
+    with open(op.join(dpath, 'data_RGCs.json'), 'r') as f:
+        rgcs_dataset = json.loads(f.read())
+
+stim = np.array(rgcs_dataset['stim'])
+stim_times = np.array(rgcs_dataset['stim_times'])
+
+n_cells = len(rgcs_dataset['spike_times'])
+dt = stim_times[1] - stim_times[0]  # time between the stimulation
+n_t = len(stim)  # total number of the stimulation
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]:
```suggestion
n_times = len(stim)  # total number of the stimulation
```

File: examples/plot_glm_retinal_ganglia.py (hunk 21 above)
Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: Yes, it just gave different scores. I will do further check on this.

**Comment by titipata** [comment]
@jasmainak I resolved most of your comments. I could change the `lg-GLM` to the following:

``` py
from sklearn.linear_model import LinearRegression

lg_model = LinearRegression()
lg_model.fit(Xdsgn, spikes_binned)

lg_model.score(Xdsgn, spikes_binned)
```

However, I think it will loosen the purpose of being a tutorial. Therefore, I just kept it as is where I change to `pinv` instead.

File: examples/plot_glm_retinal_ganglia.py (hunk 18 above)
Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: Alright, change to `GLM(distr='gaussian')` instead!

**Comment by titipata** [comment]
@jasmainak, made changes as you suggested. Let me know if it looks good to you!

File: examples/plot_glm_retinal_ganglia.py [hunk 24]
```diff
@@ -0,0 +1,259 @@
...
+#
+# Fetch the dataset. The JSON file contains the
+# following keys:  ``stim`` (binary stochastic stimulation),
+# ``stim_times`` (time of the stimulation), and
+# ``spike_times`` (recorded time of the spikes)
+
+from pyglmnet.datasets import fetch_rgc_data
+
+with TemporaryDirectory(prefix="tmp_glm-tools") as temp_dir:
+    dpath = fetch_rgc_data(dpath=temp_dir, accept_rgcs_license=True)
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: this will mean that users who download our script will not give consent to accepting the license. Unfortunately in Pyglmnet we don't have a very sophisticated download system for datasets. Ideally what you'd want to do is download the dataset in circleCI as a separate command:

```sh
- python -c "import pyglmnet; pyglmnet.datasets.fetch_rgc_data(accept_rgcs_license=True);"
```
And here in the example you would call the function with `accept_rgcs_license=False`. But it will only be invoked for users since for continuous integration, the data is already downloaded.

Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: @jasmainak I tried to follow an example from MNE (https://github.com/mne-tools/mne-python/blob/master/mne/datasets/brainstorm/bst_raw.py#L28-L33). However, I'm kinda stuck doing it. Can I ask for your help on this comment?

File: examples/plot_glm_retinal_ganglia.py [hunk 25]
```diff
@@ -0,0 +1,259 @@
...
+                       reg_lambda=1e-7)
+
+# fitting to a design matrix with spikes history
+glm_poisson_hist.fit(Xdsgn_hist, spikes_binned)
+
+# predict spike counts
+spikes_pred_poissonGLM_hist = glm_poisson_hist.predict(Xdsgn_hist)
+
+#############################################################################
+# **Putting all together**
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: @titipata why is the yellow curve a constant prediction?

Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: @jasmainak oops, sorry, I forgot to set `reg_lambda=0.0`.

File: examples/plot_glm_retinal_ganglia.py [hunk 26]
```diff
@@ -0,0 +1,264 @@
...
+# :math:`X` is the stimulation and history of stimulation
+#
+# You can simply use linear Gaussian GLM with no regularization
+# to fit as follows
+
+glm_lgGLM = GLM(distr='gaussian', reg_lambda=0.0)
+glm_lgGLM.fit(Xdsgn, spikes_binned)
+
+# predict spike counts
+spikes_pred_lgGLM_offset = glm_lgGLM.predict(Xdsgn)
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]:
```suggestion
spikes_pred_lgGLM = glm_lgGLM.predict(Xdsgn)
```

since we do `fit_intercept=True` by default, the offset terminology seems a bit overkill here?

File: examples/plot_glm_retinal_ganglia.py [hunk 27]
```diff
@@ -0,0 +1,264 @@
...
+
+########################################################
+# **Fitting and predicting with a linear-Gaussian GLM**
+#
+# For a general linear model, an observed spikes can be
+# thought of an underlying parameter
+# :math:`\beta_0, \beta` that control the spiking:
+# :math:`y = (\beta_0 + X \beta) \ + \epsilon` and
+# :math:`y \sim \text{Poiss}(\beta_0 + X \beta)` where
+# :math:`X` is the stimulation and history of stimulation
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: I wonder if we should get rid of this math here. People can refer to the tutorial if they're interested in more details.

File: examples/plot_glm_retinal_ganglia.py [hunk 28]
```diff
@@ -0,0 +1,264 @@
...
+
+########################################################
+# **Fitting and predicting with a Gaussian GLM**
+#
+# We can also assume that their is a non-linear function governing
+# the underlying the firing patterns. Concreately, we can write down as
+# :math:`y = q(\beta_0 + X \beta) + \epsilon`, and
+# :math:`y \sim \text{Poiss}(\beta_0 + X \beta)`.
+# Here we use :math:`exp()` as the inverse link function :math:`q`.
+# We can use the :class:`GLM` class to predict the parameters.
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: same here. I'd get rid of the math

File: examples/plot_glm_retinal_ganglia.py [hunk 29]
```diff
@@ -0,0 +1,264 @@
...
+# In this case, we will pick cell number 2 (ON cell).
+
+cell_idx = 2  # pick cell number 2 (ON cell)
+spike_time = rgcs_dataset['spike_times']['cell_%d' % cell_idx]
+spike_time = np.array(spike_time)
+n_spikes = len(spike_time)  # number of spikes
+
+# bin the spikes
+t_bins = np.arange(n_times + 1) * dt
+spikes_binned, _ = np.histogram(spike_time, t_bins)
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: what do you think of calling this variable `y` or at least using `y` when you do `fit` and `predict` so we stay as close to machine learning terminology as possible

File: examples/plot_glm_retinal_ganglia.py (hunk 24 above)
Comment (jasmainak) [reply examples/plot_glm_retinal_ganglia.py]: MNE uses a config system. The download path is stored in a config file. It might be too much for one PR to incorporate that as well. We can do that in a subsequent PR. Just make it `False` in the example here and use the same path that you use in `CircleCI`. Maybe `$HOME/glm-data/` ? Not ideal for everyone but we can fix that later.

File: examples/plot_glm_retinal_ganglia.py (hunk 27 above)
Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: I kinda want to keep it since it can be read through without moving out of the example.

File: examples/plot_glm_retinal_ganglia.py (hunk 26 above)
Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: Yes, let's remove the offset then! Will leave it to you @jasmainak

File: examples/plot_glm_retinal_ganglia.py [hunk 30]
```diff
@@ -0,0 +1,264 @@
...
+# create possion GLM instance
+glm_poisson = GLM(distr='poisson',
+                  alpha=0.05,
+                  learning_rate=1.0,
+                  score_metric='pseudo_R2',
+                  reg_lambda=1e-7)
+glm_poisson.fit(Xdsgn, spikes_binned)
+
+# predict spike counts
+spikes_pred_poissonGLM = glm_poisson.predict(Xdsgn)
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]:
```suggestion
ypred_poisson = glm_poisson.predict(Xdsgn)
```

GLM is common to all three methods, so it seems a bit superfluous in the name

File: examples/plot_glm_retinal_ganglia.py [hunk 31]
```diff
@@ -0,0 +1,264 @@
...
+# to fit as follows
+
+glm_lgGLM = GLM(distr='gaussian', reg_lambda=0.0)
+glm_lgGLM.fit(Xdsgn, spikes_binned)
+
+# predict spike counts
+spikes_pred_lgGLM_offset = glm_lgGLM.predict(Xdsgn)
+
+########################################################
+# **Fitting and predicting with a Gaussian GLM**
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: you mean Poisson GLM?

File: examples/plot_glm_retinal_ganglia.py [hunk 32]
```diff
@@ -0,0 +1,264 @@
...
+#
+# We can also assume that their is a non-linear function governing
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]:
```suggestion
# We can also assume that there is a non-linear function governing
```

File: examples/plot_glm_retinal_ganglia.py (hunk 27 above)
Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: Equations removed

File: examples/plot_glm_retinal_ganglia.py (hunk 29 above)
Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: made changes as you suggested.

**Comment by jasmainak** [comment]
PR looks in pretty good shape. But we need the PR on the dataset page to be merged first by @pavanramkumar 

also review is welcome @pavanramkumar

**Comment by pavanramkumar** [comment]
thanks @titipata for the really awesome initiative! i will do a review this week.

@jasmainak what is the dataset PR you are referring to? we need to be able to integrate the dataset fetcher into our example while making sure users don't get access to raw data. you may have already addressed above, in which case i'll pay attention to that aspect as well in my review.

**Comment by pavanramkumar** [comment]
> thanks @titipata for the really awesome initiative! i will do a review this week.
> 
> @jasmainak what is the dataset PR you are referring to? we need to be able to integrate the dataset fetcher into our example while making sure users don't get access to raw data. you may have already addressed above, in which case i'll pay attention to that aspect as well in my review.

merged the datasets PR; to be clear, we are ok with users having access to the data as long as we make a reproduce a dataset license statement in the README? did we explicitly get permission for the same?

**Comment by jasmainak** [comment]
@titipata you overwrote my commits and rebase :(

**Comment by titipata** [comment]
@pavanramkumar, yes, we already added the license to the dataset. However, as @jasmainak mentioned, for creating local documentation, people might accidentally get the dataset without accepting the license agreement. I will make a separate PR on that issue.

File: examples/plot_glm_retinal_ganglia.py [hunk 33]
```diff
@@ -0,0 +1,253 @@
...
+
+########################################################
+# **Fitting and predicting with a linear-Gaussian GLM**
+#
+# For a general linear model, the observed spikes can be
+# thought of an underlying parameter
+# :math:`\beta_0, \beta` that control the spiking.
+#
+# You can simply use linear Gaussian GLM with no regularization
+# to predict the parameters
```
Comment (jasmainak) [inline examples/plot_glm_retinal_ganglia.py]: but we are not predicting the `beta_0` and `beta`, no?

Comment (titipata) [reply examples/plot_glm_retinal_ganglia.py]: Oh yeah, predicting the spike counts, can you fix it for me?

**Comment by jasmainak** [comment]
It's in, congrats @titipata ! :beers:

**Comment by titipata** [comment]
will do a follow up PR soon!!
